src/core/python2js.c

```diff
@@ -11,83 +11,77 @@
 
 static PyObject* tbmod = NULL;
 
+_Py_IDENTIFIER(format_exception);
+
 static JsRef
 _python2js_unicode(PyObject* x);
 
 void
 pythonexc2js()
 {
-  PyObject* type;
-  PyObject* value;
-  PyObject* traceback;
-  int no_traceback = 0;
+  bool success = false;
+  PyObject* type = NULL;
+  PyObject* value = NULL;
+  PyObject* traceback = NULL;
+  JsRef excval = NULL;
+  PyObject* pylines = NULL;
+  PyObject* empty = NULL;
+  PyObject* pystr = NULL;
 
   PyErr_Fetch(&type, &value, &traceback);
   PyErr_NormalizeException(&type, &value, &traceback);
 
-  JsRef excval = NULL;
-  int exc;
-
   if (type == NULL || type == Py_None || value == NULL || value == Py_None) {
     excval = hiwire_string_ascii("No exception type or value");
-    PyErr_Print();
-    PyErr_Clear();
-    goto exit;
+    PySys_WriteStderr("No exception type or value\n");
+    goto finally__skip_print_tb;
+    return;
   }
 
-  if (tbmod == NULL) {
-    tbmod = PyImport_ImportModule("traceback");
-    if (tbmod == NULL) {
-      PyObject* repr = PyObject_Repr(value);
-      if (repr == NULL) {
-        excval = hiwire_string_ascii("Could not get repr for exception");
-      } else {
-        excval = _python2js_unicode(repr);
-        Py_DECREF(repr);
-      }
-      goto exit;
-    }
+  if (traceback == NULL) {
+    traceback = Py_None;
+    Py_INCREF(traceback);
   }
 
-  PyObject* format_exception;
-  if (traceback == NULL || traceback == Py_None) {
-    no_traceback = 1;
-    format_exception = PyObject_GetAttrString(tbmod, "format_exception_only");
-  } else {
-    format_exception = PyObject_GetAttrString(tbmod, "format_exception");
+  pylines = _PyObject_CallMethodIdObjArgs(
+    tbmod, &PyId_format_exception, type, value, traceback, NULL);
+  FAIL_IF_NULL(pylines);
+  empty = PyUnicode_New(0, 0);
+  FAIL_IF_NULL(empty);
+  pystr = PyUnicode_Join(empty, pylines);
+  FAIL_IF_NULL(pystr);
+  const char* pystr_utf8 = PyUnicode_AsUTF8(pystr);
+  FAIL_IF_NULL(pystr_utf8);
+  PySys_WriteStderr("Python exception:\n");
+  PySys_WriteStderr("%s\n", pystr_utf8);
+  excval = _python2js_unicode(pystr);
```

Comment:
If I understand this correctly, if this line fails, then the error indicator will be set, but no error is thrown. This does not seem ideal to me. Is there any reason this is not a `FAIL_IF_NULL(excval)` and let the `!success` logic deal with it?

Reply:
Good catch, I think that you do understand correctly and this is a logic error. Will fix!